[RN Mobile] Fix crash when merging block

Tug opened the pull request:
This PR aims to upgrade RichText on mobile native to use the latest changes from the web, mainly those made in https://github.com/WordPress/gutenberg/pull/14640

It also removes the customization we made to the rich-text library since writing in multiple formats at once is now stable on the web.

### Testing Instructions
Tested with https://github.com/wordpress-mobile/gutenberg-mobile/pull/949

etoledom committed:
Adding `onSelectionChange` from store to RichText

Tug committed:
First experiment fixing selection change handling in rich text

Tug committed:
Update readme change postcommit

nerrad commented:
Hey just a couple friendly reminders:

- For WIP it's more project friendly to use the [draft pull](https://help.github.com/en/articles/about-pull-requests#draft-pull-requests) feature in github because automatic codeowner and any requested reviewers will not get notified until the pull is published.  So you can still work on it without worrying about increasing the notification noise to potential reviewers.
- I started looking at this before noticing it was a WIP, on the surface it looks like you're not basing this branch off of a clean master because there seems to be a number of unrelated changes in the diff.  You're likely already aware but just thought I'd point out in case you weren't.

Tug committed:
Fix sequence of events on split and selection change

Tug committed:
Remove commented onTextUpdate call on enter pressed

Tug commented on packages/block-editor/src/components/rich-text/index.native.js:
```diff
@@ -103,22 +105,33 @@ export class RichText extends Component {
-			start: 0,
-			end: 0,
-			formatPlaceholder: null,
+			activeFormats: [],
```
Those updates (`formatPlaceholder`=>`activeFormats`, `this.lastContent` => `this.value`, `this.formatToValue`...) aren't required per se but they bring the code closer to the web version

Tug committed:
Do not reset activeFormats when onSelectionChange is emitted while typing

Tug commented on packages/block-editor/src/components/rich-text/index.native.js:
```diff
@@ -396,8 +445,9 @@ export class RichText extends Component {
-				this.lastContent = this.valueToFormat( linkedRecord );
-				this.props.onChange( this.lastContent );
+				this.value = this.valueToFormat( linkedRecord );
+				//this.lastEventCount = undefined;
```
Need to find out if this is required or not

Tug committed:
Update value from props

Tug committed:
Pinpoint conditions for the case where onSelectionChange is emitted without changing selection

Tug committed:
Improve isTyping detection

Tug committed:
Fix applying link format

Tug committed:
Do not try to update text from the paste handler

hypest committed:
onChange should update the state

hypest committed:
Selection change events always fire after text already changed

etoledom committed:
Fix issue where the list block was not being selected on focus.

etoledom committed:
Add isSelected prop from context to RichText

etoledom commented:
Found an issue on split (iOS and Android):

Steps to reproduce:
- Select a paragraph block in the middle of the content.
- Split the block.
- Merge back the new block.
- (Android) Split again. 
  - (iOS) Move the cursor and then Split again.
- You will see that all the content stays in the first block.

![phantom_split](https://user-images.githubusercontent.com/9772967/57483709-be572d80-72a7-11e9-9255-5b9ab23f0b69.gif)

etoledom committed:
Fix move caret to the end when merging.
This avoid moving the caret to the end when two blocks, both with content, are merged.

hypest committed:
Caret position from props, don't force to end

hypest committed:
Fix lint errors

hypest committed:
WIP: more fixes

hypest committed:
WIP: Fix list external Enter key on Android

hypest commented:
FYI, still working on some issue fixes (pushed some WIP commits to `rnmobile/fix-crash-on-merge-2-extra-fixes`).

hypest committed:
No stray onSelectionChanged events from AztecAndroid

hypest committed:
Only block html elements will be trimmed by AztecAndroid

etoledom commented on packages/block-editor/src/components/rich-text/index.native.js:
```diff
@@ -216,40 +259,55 @@ export class RichText extends Component {
+		// 	///
+		// } else {
+		// 	if ( doUpdateChild ) {
+		// 		this.lastEventCount = undefined;
```
@hypest - This is the source of the format issue on iOS. If eventCount is other than undefined, Aztec-iOS will not update its content, so we need to undefine it at some point.

etoledom replied:
I'm not sure what was the reason to do this change, but if it's working properly on Android, maybe we can modify the native iOS Aztec wrapper to work as the Android one does?

hypest replied:
Chatted over Slack and I think the culprit is to be found in the `shouldComponentUpdate` function? The assumption we can try to align to is that a "normal" prop change (initiated by https://github.com/WordPress/gutenberg/pull/15392/files/380660e7d751834ac0de4eb7f207d0c0f89fa863#diff-4828a21853e899e5a36faecfa96d83e8R272) should be enough to cause the code to try and force the changes to the Aztecs.

etoledom committed:
Fix issue on iOS where the format buttons were not working properly.
Pressing on the buttons didn't change the format of the selected text.

hypest committed:
Mirror caret state on text change

etoledom committed:
Restore iOS autoscroll while editing.

hypest committed:
Merge commit '295a381838c03ace2d239a6cbb426b1285c1cc69' into rnmobile/fix-crash-on-merge-2

hypest committed:
Cleanup

SergioEstevao commented on packages/block-editor/src/components/rich-text/index.native.js:
```diff
@@ -343,6 +391,8 @@ export class RichText extends Component {
 			onRemove( ! isReverse );
 		}
+
+		event.preventDefault();
```
Does this make sense on RN?

Tug replied:
Not really indeed.
I don't think it's doing anything at the moment (it probably could if we handled it on the native side?), I just added it because I wanted the code to get closer to the RichText web version so we can merge the two in the short term.
If you find it confusing I don't mind removing it.

SergioEstevao replied:
It's ok for me, it will actually be nice that this sent something back to the native side, and then we could decide how to handle the Enter/Delete after JS processed the event.

Tug committed:
Try a new approach to detect if the selection change event is manual or not

etoledom committed:
Fix iOS issue with formats, where pressing the format buttons sometimes won't change the format of the selected text

hypest reviewed (approved):
LGTM!

Tug reviewed (commented):
LGTM as well!

etoledom reviewed (approved):
Looks good! 🎉
